### ISSUE TITLE
Failed in training fine-grained categorization dataset CUB-200-2011

### ISSUE THREAD

**Issue by MichaelLiang12** [VisualComputingInstitute/triplet-reid#26]
I use the same triplet loss (with BatchHard, Euclidean distance  and Soft-margin) on the fine-grained categorization dataset [CUB-200-2011.](http://www.vision.caltech.edu/visipedia/CUB-200-2011.html) It aims to distinguish different species of birds (200 categorization, 5994 image in all). I know fine-grained categorization is a kind of classification task. But I want to see if it is possible to treat it as an image retrieval problem (Or Person reid).

However, when I use VGG16 (pre-trained on ImageNet) to extract features for images, and train the whole model with your triplet loss. It does not converge. All images' activations  of `conv5_3` are negative values, and the activations become 0 after the following `relu` layer. It outputs same features (the last `fc` layer) for different images at last.

I follow your instructions but use another dataset. The loss drops at first and then keep at 0.7. Nonzero triplets never decreases.

**Comment by lucasb-eyer** [comment]
Yes, that is very possible. As I write in the readme in big, bold letters:

**:boom: :fire: :exclamation: If you train on a very different dataset, don't forget to tune the learning-rate :exclamation: :fire: :boom:**

I don't write it for fun. I have trained our triplet model (with ResNet50) successfully on CUB-200-2011 (and CARS196 and Stanford Online Products, the usual three) reaching state-of-the-art, but **only when adapting the learning-rate!**

Actually, with Adam, even the "epsilon" can be an important hyper-parameter; this is also mentioned in the TensorFlow documentation of it. I recommend doing a grid-search for good hyper-parameters where the loss doesn't get stuck. (Do **not** choose by evaluating all on the test-set, that would be overfitting and cheating.) You can well do this overnight if you do only short runs of ~2000 or so updates.

The problem with CUB-200-2011 and CARS196 is that they are tiny, I find it a pity that they are still used in new publications :cry:

**Comment by maxisme** [comment]
@MichaelLiang12 did you manage to train a custom dataset which would have a loss below 0.7? I have tried many different learning rates but I have not once seen a `min loss` drop below 0.65. What does this affectively mean? "Nonzero triplets never decreases" - not quite sure what that means?

____

I am using the vgg dataset with the file structure like this:
```
class_a/file.jpg
class_b/file.jpg
class_b/file.jpg
...
```

I set the `pids, fids = [], []` like this:

```
classes = [path for path in os.listdir(DATA_DIR) if os.path.isdir(os.path.join(DATA_DIR, path))]
for c in classes:
    for file in glob.glob(DATA_DIR+c+"/*.jpg"):
        pids.append(c)
        fids.append(file)
```
where `DATA_DIR` is the directory of the vgg dataset.

**Comment by maxisme** [comment]
Yes the latter is what I am very worried about. Do you know if there any way to efficiently output an **anchor**, **positive** and **negative** at some point during training? I am new to TF. I have tried the other optimiser you mentioned in the paper no difference. I am judging that the training is working on the fact that if the min loss drops below 0.69 early on for efficiency sake is this a really poor idea?

Thank you very much for getting back to me!

**Comment by lucasb-eyer** [comment]
For completeness, I'm repeating here what I wrote in an e-mail to you. For especially difficult datasets, you really need to try many hyperparameter values to find some that converge. For example, try learning rates 1e-1, 3e-2, 1e-2, 3e-3, ..., 1e-8, each of them for 2-5k updates, and stop early for those that don't go below maybe 0.65 or so. That should be feasible in one night or so.

About outputting things, have a look at "summaries" in tensorflow, you can output even images and look at them in tensorboard.

It doesn't make sense to try optimizer settings from other papers, every model/dataset combination has their own sweet spot, so if the paper you take one from doesn't use the same model/dataset as you, it's only luck if the optimizer setting will work or not. However, RMSProp is a good one to try, too.

**Comment by maxisme** [comment]
Okay after starting the sesh I have ran:

    print(sess.run(fids))
    print(sess.run(pids))

this returns:

```
['/path/to/data-set/n006101/0320_01.jpg'
 '/path/to/data-set/n006101/0131_05.jpg'
 '/path/to/data-set/n006101/0043_01.jpg'
 '/path/to/data-set/n006101/0114_01.jpg'
 '/path/to/data-set/n004359/0174_04.jpg'
 '/path/to/data-set/n004359/0058_01.jpg'
 '/path/to/data-set/n004359/0241_01.jpg'
 '/path/to/data-set/n004359/0175_01.jpg'
 '/path/to/data-set/n007003/0417_05.jpg'
 '/path/to/data-set/n007003/0209_01.jpg'
 '/path/to/data-set/n007003/0077_01.jpg'
 '/path/to/data-set/n007003/0057_01.jpg'
 '/path/to/data-set/n000457/0203_01.jpg'
 '/path/to/data-set/n000457/0159_01.jpg'
 '/path/to/data-set/n000457/0197_01.jpg'
 '/path/to/data-set/n000457/0161_01.jpg'
 '/path/to/data-set/n000549/0363_02.jpg'
 '/path/to/data-set/n000549/0141_01.jpg'
 '/path/to/data-set/n000549/0414_01.jpg'
 '/path/to/data-set/n000549/0328_01.jpg'
 '/path/to/data-set/n006797/0147_02.jpg'
 '/path/to/data-set/n006797/0035_02.jpg'
 '/path/to/data-set/n006797/0101_02.jpg'
 '/path/to/data-set/n006797/0145_01.jpg'
 '/path/to/data-set/n001789/0210_01.jpg'
 '/path/to/data-set/n001789/0012_01.jpg'
 '/path/to/data-set/n001789/0087_01.jpg'
 '/path/to/data-set/n001789/0159_01.jpg'
 '/path/to/data-set/n000473/0074_01.jpg'
 '/path/to/data-set/n000473/0039_02.jpg'
 '/path/to/data-set/n000473/0174_01.jpg'
 '/path/to/data-set/n000473/0211_01.jpg'
 '/path/to/data-set/n000489/0008_01.jpg'
 '/path/to/data-set/n000489/0131_02.jpg'
 '/path/to/data-set/n000489/0176_01.jpg'
 '/path/to/data-set/n000489/0221_01.jpg'
 '/path/to/data-set/n002198/0159_07.jpg'
 '/path/to/data-set/n002198/0033_02.jpg'
 '/path/to/data-set/n002198/0181_03.jpg'
 '/path/to/data-set/n002198/0126_01.jpg'
 '/path/to/data-set/n000777/0445_01.jpg'
 '/path/to/data-set/n000777/0126_01.jpg'
 '/path/to/data-set/n000777/0456_04.jpg'
 '/path/to/data-set/n000777/0392_01.jpg'
 '/path/to/data-set/n007482/0196_03.jpg'
 '/path/to/data-set/n007482/0013_01.jpg'
 '/path/to/data-set/n007482/0344_01.jpg'
 '/path/to/data-set/n007482/0064_01.jpg'
 '/path/to/data-set/n005586/0061_02.jpg'
 '/path/to/data-set/n005586/0100_01.jpg'
 '/path/to/data-set/n005586/0144_01.jpg'
 '/path/to/data-set/n005586/0382_01.jpg'
 '/path/to/data-set/n000944/0317_01.jpg'
 '/path/to/data-set/n000944/0144_01.jpg'
 '/path/to/data-set/n000944/0469_01.jpg'
 '/path/to/data-set/n000944/0030_01.jpg'
 '/path/to/data-set/n003644/0695_01.jpg'
 '/path/to/data-set/n003644/0104_02.jpg'
 '/path/to/data-set/n003644/0032_01.jpg'
 '/path/to/data-set/n003644/0131_01.jpg'
 '/path/to/data-set/n006191/0241_01.jpg'
 '/path/to/data-set/n006191/0186_03.jpg'
 '/path/to/data-set/n006191/0073_01.jpg'
 '/path/to/data-set/n006191/0157_05.jpg'
 '/path/to/data-set/n004641/0269_01.jpg'
 '/path/to/data-set/n004641/0030_01.jpg'
 '/path/to/data-set/n004641/0179_01.jpg'
 '/path/to/data-set/n004641/0132_01.jpg'
 '/path/to/data-set/n000881/0086_01.jpg'
 '/path/to/data-set/n000881/0351_03.jpg'
 '/path/to/data-set/n000881/0233_01.jpg'
 '/path/to/data-set/n000881/0130_01.jpg']
['n000203' 'n000203' 'n000203' 'n000203' 'n009265' 'n009265' 'n009265'
 'n009265' 'n006279' 'n006279' 'n006279' 'n006279' 'n005480' 'n005480'
 'n005480' 'n005480' 'n005396' 'n005396' 'n005396' 'n005396' 'n007609'
 'n007609' 'n007609' 'n007609' 'n002699' 'n002699' 'n002699' 'n002699'
 'n008955' 'n008955' 'n008955' 'n008955' 'n000885' 'n000885' 'n000885'
 'n000885' 'n007587' 'n007587' 'n007587' 'n007587' 'n008725' 'n008725'
 'n008725' 'n008725' 'n006369' 'n006369' 'n006369' 'n006369' 'n008052'
 'n008052' 'n008052' 'n008052' 'n000116' 'n000116' 'n000116' 'n000116'
 'n008270' 'n008270' 'n008270' 'n008270' 'n000668' 'n000668' 'n000668'
 'n000668' 'n006747' 'n006747' 'n006747' 'n006747' 'n002827' 'n002827'
 'n002827' 'n002827']
```
Am I correct in thinking things have gone terribly wrong hence the first two files are not class `n000203` etc...?

**Comment by lucasb-eyer** [comment]
Yes, good find, it seems things have gone wrong in the dataset preparation somewhere!

**Comment by lucasb-eyer** [comment]
Could you please post again your comments here, but in a new issue? I'd like to delete them from this issue as they are not related to the original issue and distract from it.

**Comment by maxisme** [comment]
Is this okay? https://github.com/VisualComputingInstitute/triplet-reid/issues/30

**Comment by lucasb-eyer** [comment]
Yup, let's do our other messages too, and then I'll remove them here, thank you! :smile:
